src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -427,74 +419,81 @@ HeapWord* ParallelScavengeHeap::mem_allocate_old_gen(size_t size) {
     return allocate_old_gen_and_record(size);
   }
 
-  // If a "death march" is in progress, allocate from the old gen a limited
-  // number of times before doing a GC.
-  if (_death_march_count > 0) {
-    if (_death_march_count < 64) {
-      ++_death_march_count;
-      return allocate_old_gen_and_record(size);
-    } else {
-      _death_march_count = 0;
-    }
-  }
   return nullptr;
 }
 
 void ParallelScavengeHeap::do_full_collection(bool clear_all_soft_refs) {
-  PSParallelCompact::invoke(clear_all_soft_refs);
-}
-
-// Failed allocation policy. Must be called from the VM thread, and
-// only at a safepoint! Note that this method has policy for allocation
-// flow, and NOT collection policy. So we do not check for gc collection
-// time over limit here, that is the responsibility of the heap specific
-// collection methods. This method decides where to attempt allocations,
-// and when to attempt collections, but no collection specific policy.
-HeapWord* ParallelScavengeHeap::failed_mem_allocate(size_t size) {
-  assert(SafepointSynchronize::is_at_safepoint(), "should be at safepoint");
-  assert(Thread::current() == (Thread*)VMThread::vm_thread(), "should be in vm thread");
-  assert(!is_stw_gc_active(), "not reentrant");
-  assert(!Heap_lock->owned_by_self(), "this thread should not own the Heap_lock");
+  if (GCLocker::check_active_before_gc()) {
+    return;
+  }
+  do_full_collection_no_gc_locker(clear_all_soft_refs);
+}
 
-  // We assume that allocation in eden will fail unless we collect.
+void ParallelScavengeHeap::do_full_collection_no_gc_locker(bool clear_all_soft_refs) {
+  bool maximum_compaction = clear_all_soft_refs;
+  PSParallelCompact::invoke(maximum_compaction);
+}
 
-  // First level allocation failure, scavenge and allocate in young gen.
-  GCCauseSetter gccs(this, GCCause::_allocation_failure);
-  const bool invoked_full_gc = PSScavenge::invoke();
-  HeapWord* result = young_gen()->allocate(size);
+HeapWord* ParallelScavengeHeap::expand_heap_and_allocate(size_t size, bool is_tlab) {
+  HeapWord* result = nullptr;
 
-  // Second level allocation failure.
-  //   Mark sweep and allocate in young generation.
-  if (result == nullptr && !invoked_full_gc) {
-    do_full_collection(false);
-    result = young_gen()->allocate(size);
+  result = young_gen()->allocate(size);
+  if (result == nullptr && !is_tlab) {
+    // auto expand inside
+    result = old_gen()->allocate(size);
```

Comment:
If we expand the generation in the method `PSOldGen::allocate`. I think it is good to rename the method to `expand_and_allocate` (just like the SerialGC). It is better to be polished at a followup issue.